### ISSUE TITLE
pubsub: sub.Receive does not respect context deadlines when messages are not ACKed

### ISSUE THREAD

**Issue by nathanperkins** [googleapis/google-cloud-go#11070]
## Client

cloud.google.com/go/pubsub v1.45.1

## Environment

- go 1.23
- pubsub emulator 499.0.0: `docker run --rm -p 8113:8113 gcr.io/google.com/cloudsdktool/google-cloud-cli:499.0.0-emulators gcloud beta emulators pubsub start --host-port=0.0.0.0:8113`
- var `PUBSUB_EMULATOR_HOST=localhost:8113`

## Code and Dependencies

```go
func TestPubSub(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	t.Cleanup(cancel)

	ps, err := pubsub.NewClient(ctx, "project-id")
	if err != nil {
		t.Fatalf("Failed to create pubsub client: %s", err)
	}

	top, err := ps.CreateTopic(ctx, "topic-id")
	if err != nil {
		t.Fatalf("Failed to create pubsub topic: %s", err)
	}
	t.Logf("Created topic: %s", top.String())

	subCfg := pubsub.SubscriptionConfig{
		Topic: top,
	}
	sub, err := ps.CreateSubscription(ctx, "subscription-id", subCfg)
	if err != nil {
		t.Fatalf("Failed to create pubsub subscription: %s", err)
	}
	t.Logf("Created subscription: %s", sub.String())

	res := top.Publish(ctx, &pubsub.Message{
		Data: []byte("foo"),
	})
	if _, err := res.Get(ctx); err != nil {
		t.Fatalf("Failed to publish message: %s", err)
	}
	t.Logf("Published 1 message")

	var msgs []*pubsub.Message
	recv := func(_ context.Context, msg *pubsub.Message) {
		t.Logf("Got msg: %v", msg.Data)
		msgs = append(msgs, msg)
		// msg.Ack()
	}

	// cancel()
	if err := sub.Receive(ctx, recv); err != nil {
		t.Fatalf("Failed to receive pubsub messages: %s", err)
	}
	t.Logf("Received %d pubsub messages", len(msgs))
}

```

## Expected behavior

sub.Receive should immediately unblock due to the context already being cancelled and we should see t.Fatalf log due to context err.

This seems to be a bug in the client related to context deadlines when messages are not ACKed. You can see in the code, I've commented out `msg.Ack()` within the recv function, and `cancel()` right before `sub.Receive(...)`. If you uncomment either one, the code no longer blocks.

## Actual behavior

sub.Receive blocks forever until the test times out.

**Comment by nathanperkins** [comment]
Based on further investigation, we figured out that it's actually not respecting the context deadline when the recv function doesn't ACK messages. Cancellation works fine.

**Comment by hongalex** [comment]
The cancellation is possibly due to [this line](https://github.com/googleapis/google-cloud-go/blob/e1f67b11c82678cb410dc16e9bb05e183ad7877e/pubsub/iterator.go#L282) that only checks for `context.Canceled` and not `context.DeadlineExceeded`

**Comment by nathanperkins** [comment]
There's another line I found with [this query](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/googleapis/google-cloud-go%24+f:pubsub/+%22context.Canceled%22+-f:_test.go&patternType=keyword&sm=0)

https://github.com/googleapis/google-cloud-go/blob/e1f67b11c82678cb410dc16e9bb05e183ad7877e/pubsub/iterator.go#L409

```
// Get messages using the Pull RPC.
// This may block indefinitely. It may also return zero messages, after some time waiting.
func (it *messageIterator) pullMessages(maxToPull int32) ([]*pb.ReceivedMessage, error) {
	// Use it.ctx as the RPC context, so that if the iterator is stopped, the call
	// will return immediately.
	res, err := it.subc.Pull(it.ctx, &pb.PullRequest{
		Subscription: it.subName,
		MaxMessages:  maxToPull,
	}, gax.WithGRPCOptions(grpc.MaxCallRecvMsgSize(maxSendRecvBytes)))
	switch {
	case err == context.Canceled:
		return nil, nil
	case status.Code(err) == codes.Canceled: // <<<<<<
		return nil, nil
	case err != nil:
		return nil, err
	default:
		return res.ReceivedMessages, nil
	}
}
```

**Comment by nathanperkins** [comment]
Thanks for looking into this with me :)

Given that this was only happening since I was not ACKing the messages, this is no longer an issue for me. You could consider it a lower priority or even `wontfix` since the docs say we must ACK or NACK the messages within the recv function.

**Comment by hongalex** [comment]
After some more investigation, the issue isn't a difference between `context.Cancelled` versus `context.DeadlineExceeded` errors. The current behavior of `Receive` is to not return until ALL outstanding messages are either acked, nacked, or expired. In the repro code above, initiating a receive that later times out still results in a message pulled in. Because the code doesn't call `message.Ack`, the library holds onto the message for the full default `MaxExtension` of 10 minutes, which blocks `Receive` from returning. When cancelling the context before passing into `Receive`, the message is never pulled in the first place, allowing the function to return early.

@nathanperkins had the feedback that it is surprising that context cancellation in Go doesn't result in the function returning right away (or shortly after), and is blocked on messages being processed. Instead, the assumption "is that if i cancel a context, all the operations using the context will unblock pretty quickly".

While we can't change the default behavior due to this being a breaking change, we plan to introduce options to configure the shutdown behavior of `Receive` when the context canceled (and deadline exceeded) signal is sent. Instead of blocking on all messages being processed, we could (for example) introduce an option that makes the library `nack` all outstanding messages.

Closing this issue since the original issue isn't a bug with the library. Thanks for your help and feedback!
